src/systolic_array/unit_test/test_tensor_iterator.cpp

```diff
@@ -0,0 +1,149 @@
+#include <iostream>
+#include <vector>
+#include <memory>
+
+#include "aladdin/unit-test/catch.hpp"
+#include "systolic_array/tensor.h"
+
+using namespace systolic;
+
+// Compute the reference results for region-based iterator operation. It's slow
+// as it literally converts the region into single step movements, but it gives
+// us the correct results.
+std::vector<int> computeReferenceResults(const TensorIndexIterator& iter,
+                                         const std::vector<int>& region) {
+  std::unique_ptr<TensorIndexIterator> refIter(iter.clone());
+  int numSingleSteps = 0;
+  int stride = 1;
+  for (int i = region.size() - 1; i >= 0; i--) {
+    numSingleSteps += region[i] * stride;
+    stride *= iter.getDimSize(i);
+  }
+  for (int i = 0; i < numSingleSteps; i++)
+    ++(*refIter);
+  return refIter->getIndices();
+}
+
+void doRegionTest(TensorIndexIterator& iter, const std::vector<int>& region) {
+  std::vector<int> refResults(computeReferenceResults(iter, region));
+  iter += region;
+  REQUIRE(iter.getIndices() == refResults);
+}
+
+SCENARIO("Test tensor index iterator", "[tensor-index]") {
+  GIVEN("Test tensor shape without paddings.") {
+    TensorShape shape({ 4, 8, 8, 16 });
+    TensorIndexIterator iter(shape);
+    WHEN("Test single step advance.") {
+      THEN("Advance the iterator by 1 step.") {
+        ++iter;
+        REQUIRE(int(iter) == 1);
+      }
+      THEN("Advance the iterator by a random number of single steps.") {
+        for (int i = 0; i < 1234; i++)
+          ++iter;
+        REQUIRE(int(iter) == 1234);
+      }
+    }
+    WHEN("Test regin-based advance.") {
```

Comment:
Nit: the usage of behavior-driven tests is fine, but the phrasing here is confusing. Example: you write: "Given test tensor shape without paddings, when test region-based advance, then advance the iterator by regions that don't go out of dimension bounds". This is not a logical given-when-then phrase. Instead, it should be like "given a test tensor shape without paddings, when advance the iterator by regions that do out of bounds, then [insert expected result, like "iterator returns error" or "iterator overflows" or "iterator does something correctly"]". As it stands, I don't know what the expected result is supposed to be.

If the test doesn't naturally follow this sort of logic, then just nest SECTION instead.

Also typo: 'regin-based advance'

Reply:
I agree. I wasn't so sure with the bahavior-driven semantics :) Changed to what you suggested and fixed the typos.